pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +143,309 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
+				return entry, nil
+			}
+		}
 	}
-	nodePlan := plan.NodePlan{
-		Instructions: []plan.OneTimeInstruction{encryptionKeyRotationRestartInstruction(cp, leader, phase)},
+
+	machines := collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)
+	if len(machines) == 0 {
+		return nil, p.encryptionKeyRotationFailed(cp)
 	}
-	err = assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
-	if err != nil {
-		return p.enqueueIfErrWaiting(cp, err)
+	if cp.Status.RotateEncryptionKeysPhase != rkev1.RotateEncryptionKeysPhaseRestartNodes {
+		return nil, p.restartEncryptionKeyRotation(cp)
 	}
 
-	err = p.encryptionKeyRotationUpdateControlPlanePhase(cp, rkev1.RotateEncryptionKeysPhaseVerifyLeaderStatus)
+	cp.Annotations[encryptionKeyRotationLeaderAnnotation] = machines[0].Machine.Name
+	_, err := p.rkeControlPlanes.Update(cp)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	return generic.ErrSkip
+	// a new leader was chosen, must restart
+	return nil, generic.ErrSkip
 }
 
-func (p *Planner) encryptionKeyRotationVerifyLeaderStatus(cp *rkev1.RKEControlPlane, leader *planEntry) error {
-	return p.encryptionKeyRotationGetLeaderStatus(cp, leader, rkev1.RotateEncryptionKeysPhaseRestartFollowers)
+// encryptionKeyRotationIsSuitableControlPlane ensures that a control plane node has not been deleted and has a valid
+// node associated with it.
+func encryptionKeyRotationIsSuitableControlPlane(entry *planEntry) bool {
+	return isControlPlane(entry) && entry.Machine.DeletionTimestamp == nil && entry.Machine.Status.NodeRef != nil
 }
 
-// encryptionKeyRotationRestartFollowers restarts each follower, and ensures that the status matches the leader node.
-// If this is the first time this function has been run this generation, the generation status will be updated, however
-// the phase will not be set to done unless the controlplane secrets-encrypt status is uniformly reencrypt_finished, as
-// subsequent runs will begin in the reencrypt_finished phase, and require the applyLeader phase to be reached at least
-// once before completion.
-func (p *Planner) encryptionKeyRotationRestartFollowers(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
-	leaderPhase, err := encryptionKeyRotationPhaseFromPeriodic(leader)
+// encryptionKeyRotationRestartNodes will restart the leader to ensure that the current encryption key rotation status
+// is persisted to the datastore. With a single node control plane, the apiserver container needs to be restarted,
+// however with an HA control plane, the leader must first be restarted then all the followers, as the followers will
+// update their status when reading from the datastore on startup of the rke2/k3s services.
+func (p *Planner) encryptionKeyRotationRestartNodes(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	if rke2.GetRuntime(cp.Spec.KubernetesVersion) == rke2.RuntimeRKE2 &&
```

Comment:
Can we throw this logic under a switch statement between RKE2 and K3s to help ensure that we are more clear about this? Bonus points if we just throw RKE2/K3s under their own functions and have a master function that switches between them.